### ISSUE TITLE
empty array result after patch method

### ISSUE THREAD

**Issue by SeaDog505** [feathersjs/feathers#1221]
I use below code in my app.

```
app.service('customers').patch(null, { $set: { "active": false, "services.$": data }}, { query: { _id: Obj._id, "services.active": true } }).then(function(result){

console.log(result);

});
```

after call this method the result is works in mongoDB but the callback is empty array.
Why I receive empty array but not the document with the corectly ID?

**Comment by daffl** [comment]
If you use `Obj._id` instead of `null` it will probably work:

```js
app.service('customers').patch(Obj._id, { $set: { "active": false, "services.$": data }}, { query: { "services.active": true } }).then(function(result){
  console.log(result);
});
```

**Comment by SeaDog505** [comment]
it does not really work. In mongoDB the result is good but I receive an error:
`error: NotFound: No record found for id '5c6c3d7f0f58942d1de7e67c'`

**Comment by EliSadaka** [comment]
I'm experiencing the same issue. Was a solution found for this yet?

**Comment by EliSadaka** [comment]
I think I figured out what the issue is. If the query includes one of the fields that is being patched, it'll update the database correctly but still return the error `No record found for id ${id}`. I'm experiencing this with the Mongoose adapter but it could be an issue affecting other adapters as well. @daffl Do you know if this can be fixed?

**Comment by davidb1** [comment]
I'm getting this too, the callback returns an empty array but the objects get created

**Comment by RubyRubenstahl** [comment]
I've been having the same issue. 

I was able to trace it down to the
```JS
    // NOTE (EK): We need this shitty hack because update doesn't
    // return a promise properly when runValidators is true. WTF!
````
section of the  `service.js` file in `feathers-mongoose`. 

It seems that when an item is queried for and changed in the `data` parameter, the query is re-used to find the items again after the patch is completed. Because the query is looking for values that were changed in the patch, the query fails to find any. 

I've forked it and written a test that looks like this, which fails. 
```JS
    it('returns correct result when data includes queried properties', async () => {
      const data = await pets.patch(null, { name: 'Spot' }, { query: { name: 'Rufus' } });
      expect(data[0].name).to.equal('Spot');
    });
```

I then modified the following:
```JS
        const findParams = idList.length ? Object.assign({}, params, {
          paginate: false,
          query: Object.assign({
            [this.id]: { $in: idList }
          }, params.query)
        }) : Object.assign({}, params, {
          paginate: false
        });
```

To search through the query and replace any query properties included in the patch data with the patch data value so that it will find them in the `findParams` like so:
```JS 
      const findQuery = !params.query ? null : Object.keys(params.query)
        .reduce((query, key) => {
          query[key] = data[key] === undefined ? params.query[key] : data[key];
          return query;
        }, {});

      return ids.then(idList => {
        // Create a new query that re-queries all ids that
        // were originally changed
        const findParams = idList.length ? Object.assign({}, params, {
          paginate: false,
          query: Object.assign({
            [this.id]: { $in: idList }
          }, findQuery)
        }) : Object.assign({}, params, {
          paginate: false,
          query: findQuery
        });
```

This made my test pass without breaking anything new (for some reason I had about 20 tests that were failing before I started editing anything). 

I'm hesitant to submit it as a PR, as it's a hack on top of a hack, but I figured this might be helpful in finding a more permanent solution. 

@daffl , any thoughts? I'd be glad to put a bit more work into it and submit a PR with some direction.

**Comment by daffl** [comment]
This should have been fixed a while ago. Which version of `feathers-mongoose` are you using (current is `v7.3.2`)

**Comment by RubyRubenstahl** [comment]
I started with 6.3.0 (Installed by @feathers-plus/cli), but forked v7.3.2 to work on it. 

I added a few more checks to the test to make sure that I'm at least getting back an array:
```JS
    it('returns correct result when data includes queried properties', async () => {
      const data = await pets.patch(null, { name: 'Spot' }, { query: { name: 'Rufus' } });
      expect(data).to.be.an('array');
      expect(data.length).to.equal(1);
      expect(data[0].name).to.equal('Spot');
    });
```

When I run the test without any change to code, here's what I receive from mocha:
```
1) Feathers Mongoose Service
       Common functionality
         returns correct result when data includes queried properties:

      AssertionError: expected 0 to equal 1
      + expected - actual

      -0
      +1
```

**Comment by daffl** [comment]
Can you open a PR or issue with this test in [feathers-mongoose](https://github.com/feathersjs-ecosystem/feathers-mongoose)? Closing this issue to move the discussion over there.

**Comment by RubyRubenstahl** [comment]
Will do, I'll put it up when I have a free moment.

**Comment by matiaslopezd** [comment]
Hi @daffl , I have the same problem! I use the `v7.3.2` 

I use other way to patch multiples records:

```js
return new Promise((resolve, reject) => {
  const model = app.service('users').Model;
  model.updateMany(
  {
    // Here your filter/query
  },
  {
    $set: {
      // Here fields to patch
    }
  }
  ).then((r) => resolve(r)).catch((e) => reject(e));
});
```

**Comment by RubyRubenstahl** [comment]
I'll be submitting the info daffl asked for as a feathers-mongoose issue later today. I've been swamped with a project that I just finished yesterday and haven't had time. 

FYI, as a workaround if needed, you can patch the do a find with the same query afterword. The patch works properly, it just doesn't return afterward.

**Comment by daffl** [comment]
Fixed in feathers-mongoose v8.0.0
